pkg/unifiedpush/client.go

```diff
@@ -0,0 +1,197 @@
+package unifiedpush
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	pushv1alpha1 "github.com/aerogear/unifiedpush-operator/pkg/apis/push/v1alpha1"
+	"github.com/pkg/errors"
+	"io/ioutil"
+	"k8s.io/apimachinery/pkg/runtime"
+	"mime/multipart"
+	"net/http"
+	"strconv"
+)
+
+// variant is an internal base type
+type variant struct {
+	Name        string `json:"name"`
+	Description string `json:"description"`
+	VariantId   string `json:"variantID"`
+	Secret      string `json:"secret"`
+}
+
+// upsAndroidVariant is an internal struct used for convenient JSON
+// unmarshalling of the response received from UPS
+type upsAndroidVariant struct {
+	ProjectNumber string `json:"projectNumber"`
+	GoogleKey     string `json:"googleKey"`
+	variant
+}
+
+// upsIOSVariant is an internal struct used for convenient JSON
+// unmarshalling of the response received from UPS
+type upsIOSVariant struct {
+	Certificate []byte `json:"certificate"`
+	Passphrase  string `json:"passphrase"`
+	Production  bool   `json:"production"`
+	variant
+}
+
+// pushApplication is an internal struct used for convenient JSON
+// unmarshalling of the response received from UPS
+type pushApplication struct {
+	ApplicationId string `json:applicationId`
+	MasterSecret  string `json:masterSecret`
+}
+
+// UnifiedpushClient is a client to enable easy interaction with a UPS
+// server
+type UnifiedpushClient struct {
+	url string
+}
+
+// CreateApplication creates an application in UPS
+func (c UnifiedpushClient) CreateApplication(p *pushv1alpha1.PushApplication) (string, string, error) {
+	url := fmt.Sprintf("%s/rest/applications/", c.url)
+
+	params := map[string]string{
+		"description": p.Spec.Description,
+	}
+
+	payload, err := json.Marshal(params)
+	if err != nil {
+		return "", "", errors.Wrap(err, "Failed to marshal push application params to json")
+	}
+
+	req, err := http.NewRequest(http.MethodPost, url, bytes.NewBuffer(payload))
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("Accept", "application/json")
+
+	resp, err := doUPSRequest(req)
+	if err != nil {
+		return "", "", err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != 201 {
```

Comment:
I tried to pull as much common stuff as possible out into that `doUPSRequest` function that's called above, but I'm not sure how much makes sense.

We could consider putting this check into it, but I'm not sure if _all_ requests would expect a `201` status code response (the current "create" ones do, but we could add more read/update/delete ones).

I think putting this (or some more general check, such as "make sure the statuscode is less than 400") would allow us to pull more stuff into it, such as the reading (and closing) of the response body, and just return the byte slice from it, that we can just `json.unmarshal` here.

Again, I don't know if that's better or worse...opinions welcome!

Reply:
IMO let's leave it that way for now. We can change later if needed.